### PR TITLE
ufs: added mostly working unix filesystem server session implementation

### PR DESCRIPTION
**Pull request by progrium**
This adds a real implementation of 9ps that doesn't depend on another project. This implementation was loosely based on the go9p project [ufs implementation](https://github.com/mortdeus/go9p/tree/master/srv/examples/ufs). It's not entirely complete, but enough to use 9pr directly with 9ps with a real filesystem.

Signed-off-by: Jeff Lindsay <progrium@gmail.com>

### CODE REVIEW

File: ufs/session.go [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+	}
+
+	return ref, nil
+}
+
+func (sess *session) newRef(fid p9p.Fid, path string) (*FileRef, error) {
+	sess.Lock()
+	defer sess.Unlock()
+
+	if fid == 0 {
```
Comment (titanous) [inline ufs/session.go]: 0 is a valid fid (the Linux kernel uses it as its first fid for example). This should check against `p9p.NOFID`.

File: ufs/session.go [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+		rootRef: &FileRef{Path: root},
+		refs:    make(map[p9p.Fid]*FileRef),
+	}, nil
+}
+
+func (sess *session) getRef(fid p9p.Fid) (*FileRef, error) {
+	sess.Lock()
+	defer sess.Unlock()
+
+	if fid == 0 {
```
Comment (titanous) [inline ufs/session.go]: Ditto.

File: ufs/session.go [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+
+	if err != nil {
+		return p9p.Qid{}, 0, err
+	}
+
+	ref.Lock()
+	defer ref.Unlock()
+	ref.Path = newpath
+	ref.File = file
+	if err := ref.Stat(); err != nil {
```
Comment (titanous) [inline ufs/session.go]: This will deadlock, as `ref.Stat` attempts to lock the ref as well.

**Comment by titanous** [comment]
I was able to get this working with the Linux kernel 9p client with the commits from #39 and progrium/go-p9p#1.

**Comment by progrium** [comment]
I believe all commits have DCO now.

**Comment by progrium** [comment]
Rebased with #39 since it's required for a working, mountable implementation. Also happened to rebase with #40 because it just makes sense. If for some reason these aren't desirable I can recreate this PR, but I figured these all make sense together.

File: ufs/session.go [hunk 4]
```diff
@@ -0,0 +1,357 @@
+package ufs
```
Comment (stevvooe) [inline ufs/session.go]: What does `ufs` stand for? Is this just from the other project?

Comment (progrium) [reply ufs/session.go]: Unix filesystem

File: ufs/session.go [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+
+	if dir.Length != ^uint64(0) {
+		if err := os.Truncate(ref.Path, int64(dir.Length)); err != nil {
+			return err
+		}
+	}
+
+	// If either mtime or atime need to be changed, then
+	// we must change both.
+	//if dir.ModTime != time.Time{} || dir.AccessTime != ^uint32(0) {
```
Comment (stevvooe) [inline ufs/session.go]: I'm going to merge this, but let's get this updated in a future PR.

**Comment by stevvooe** [comment]
LGTM

**Pull request merged**